src/Components/Components/src/Routing/RouteTableFactory.cs

```diff
@@ -121,62 +120,84 @@ internal static int RouteComparison(RouteEntry x, RouteEntry y)
+            var minSegments = Math.Min(xTemplate.Segments.Length, yTemplate.Segments.Length);
+            var currentResult = 0;
+            for (var i = 0; i < minSegments; i++)
             {
-                return xTemplate.Segments.Length < y.Template.Segments.Length ? -1 : 1;
+                var xSegment = xTemplate.Segments[i];
+                var ySegment = yTemplate.Segments[i];
+
+                var xRank = GetRank(xSegment);
+                var yRank = GetRank(ySegment);
+
+                currentResult = xRank.CompareTo(yRank);
+
+                // If they are both literals we can disambiguate
+                if ((xRank, yRank) == (0, 0))
+                {
+                    currentResult = StringComparer.OrdinalIgnoreCase.Compare(xSegment.Value, ySegment.Value);
+                }
+
+                if (currentResult != 0)
+                {
+                    break;
+                }
             }
-            else
+
+            if (currentResult == 0)
             {
-                for (var i = 0; i < xTemplate.Segments.Length; i++)
-                {
-                    var xSegment = xTemplate.Segments[i];
-                    var ySegment = yTemplate.Segments[i];
-                    if (!xSegment.IsParameter && ySegment.IsParameter)
-                    {
-                        return -1;
-                    }
-                    if (xSegment.IsParameter && !ySegment.IsParameter)
-                    {
-                        return 1;
-                    }
+                currentResult = xTemplate.Segments.Length.CompareTo(yTemplate.Segments.Length);
 
-                    if (xSegment.IsParameter)
-                    {
-                        // Always favor non-optional parameters over optional ones
-                        if (!xSegment.IsOptional && ySegment.IsOptional)
-                        {
-                            return -1;
-                        }
-
-                        if (xSegment.IsOptional && !ySegment.IsOptional)
-                        {
-                            return 1;
-                        }
-
-                        if (xSegment.Constraints.Length > ySegment.Constraints.Length)
-                        {
-                            return -1;
-                        }
-                        else if (xSegment.Constraints.Length < ySegment.Constraints.Length)
-                        {
-                            return 1;
-                        }
-                    }
-                    else
-                    {
-                        var comparison = string.Compare(xSegment.Value, ySegment.Value, StringComparison.OrdinalIgnoreCase);
-                        if (comparison != 0)
-                        {
-                            return comparison;
-                        }
-                    }
+                if ((currentResult == -1 && HasRemainingOptionalSegments(yTemplate, minSegments + 1)) ||
```

Comment:
Nit: Would be helpful from a readability perspective to set these booleans to variables, something like `shorterWithOptionals` etc.

Reply:
hmm, I think it might be best to put an explaining comment rather than the explaining variable to explain why we are doing these checks, so I'll add that instead.

Reply:
This got removed because length still matters, even for optional segments. So routes are only ambiguous when the only difference is the optionality of the parameters, which was already covered by this.